src/app/ReadClient.cpp

```diff
@@ -77,95 +78,112 @@ void ReadClient::MoveToState(const ClientState aTargetState)
+                .EventId(eventPath.mEventId)
+                .EndpointId(eventPath.mEndpointId)
+                .ClusterId(eventPath.mClusterId)
+                .EndOfEventPath();
+            SuccessOrExit(err = eventPathBuilder.GetError());
+        }
 
-            eventPathListBuilder.EndOfEventPathList();
-            SuccessOrExit(err = eventPathListBuilder.GetError());
+        eventPathListBuilder.EndOfEventPathList();
+        SuccessOrExit(err = eventPathListBuilder.GetError());
 
-            if (aEventNumber != 0)
-            {
-                // EventNumber is optional
-                request.EventNumber(aEventNumber);
-            }
+        if (aEventNumber != 0)
+        {
+            // EventNumber is optional
+            request.EventNumber(aEventNumber);
         }
+    }
 
-        if (aAttributePathParamsListSize != 0 && apAttributePathParamsList != nullptr)
+    if (aAttributePathParamsListSize != 0 && apAttributePathParamsList != nullptr)
+    {
+        AttributePathList::Builder attributePathListBuilder = request.CreateAttributePathListBuilder();
+        SuccessOrExit(err = attributePathListBuilder.GetError());
+        for (size_t index = 0; index < aAttributePathParamsListSize; index++)
         {
-            AttributePathList::Builder attributePathListBuilder = request.CreateAttributePathListBuilder();
-            SuccessOrExit(err = attributePathListBuilder.GetError());
-            for (size_t index = 0; index < aAttributePathParamsListSize; index++)
+            AttributePath::Builder attributePathBuilder = attributePathListBuilder.CreateAttributePathBuilder();
+            attributePathBuilder.NodeId(apAttributePathParamsList[index].mNodeId)
+                .EndpointId(apAttributePathParamsList[index].mEndpointId)
+                .ClusterId(apAttributePathParamsList[index].mClusterId);
+
+            AttributePathSelector * selector = apAttributePathParamsList[index].mpSelector;
+
+            while (selector->mpNext != nullptr)
             {
-                AttributePath::Builder attributePathBuilder = attributePathListBuilder.CreateAttributePathBuilder();
-                attributePathBuilder.NodeId(apAttributePathParamsList[index].mNodeId)
-                    .EndpointId(apAttributePathParamsList[index].mEndpointId)
-                    .ClusterId(apAttributePathParamsList[index].mClusterId);
-                if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kFieldIdValid)
+                if (selector->mFlag == AttributePathSelectorFlag::kFieldIdValid)
                 {
-                    attributePathBuilder.FieldId(apAttributePathParamsList[index].mFieldId);
+                    attributePathBuilder.FieldId(selector->mFieldId);
                 }
-                else if (apAttributePathParamsList[index].mFlags == AttributePathFlags::kListIndexValid)
+                else if (selector->mFlag == AttributePathSelectorFlag::kListIndexValid)
```

Comment:
The else-if does not detect if several fields are set at the same time.

Please either chang that a single flag is set or accept both flags to be set. Can we have a unit test for this?